[ETHEREUM-CONTRACTS|SDK-CORE] 1544 new flow operator updates

0xdavinchee opened the pull request:
## Ethereum Contracts
- `increaseFlowAllowanceWithPermissions` and `decreaseFlowAllowanceWithPermissions` added to `ConstantFlowAgreementV1.sol`
- Slight refactoring of validation logic, **modifying existing deployed code path** (see `_validateUpdateFlowOperatorDataInput` function)

## SDK-Core
- `increaseFlowAllowanceWithPermissions` and `decreaseFlowAllowanceWithPermissions` added to `ConstantFlowAgreementV1` class

0xdavinchee committed:
add new update functions

- add increaseFlowAllowanceWithPermissions and decreaseFlowAllowanceWithPermissions
- refactor code slightly to use reuse validation logic for
- add to SuperTokenV1Library
- modify default tdd dev script + update readme

0xdavinchee committed:
add happy path tests

0xdavinchee committed:
add function to SDK

0xdavinchee committed:
Update CHANGELOG.md

kasparkallas reviewed (changes requested):
The increasing and decreasing of flow operator permissions should also function delta-based, not snapshot-based.

For example, if I increase a permission by 1 (Create) then it should never remove other permissions (Delete or Update).

Some other examples:
1. If my permissions are 1 (Create) and I increase by 1 then it stays at 1.
2. If my permissions are 1 (Create) and I increase by 2 (Update) then it becomes 3 (Create or Update).
3. If my permissions are 5 (Create or Delete) and I increase by 1 (Create) then it stays 5.
4. If my permissions are 1 (Create) and I decrease by 5 (Create or Delete) then it should become 0.
5. If My permissions are 1 (Create) and I decrease by 6 (Delete or Update) then it should stay 1.

Link to permission values for quick reference: https://docs.superfluid.finance/superfluid/developers/constant-flow-agreement-cfa/cfa-access-control-list-acl#permissions-parameter

0xdavinchee committed:
delta permissions instead of set

- delta the permissions instead of setting it and apply across sdk tests, etc.
- fix coverage as well

kasparkallas commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -934,6 +910,120 @@ contract ConstantFlowAgreementV1 is
+    ) internal returns (ISuperfluid.Context memory currentContext) {
+        if (!FlowOperatorDefinitions.isPermissionsClean(permissions)) revert CFA_ACL_UNCLEAN_PERMISSIONS();
+        // [SECURITY] NOTE: we are holding the assumption here that ctx is correct and we validate it with
+        // authorizeTokenAccess:
```
Nitpick: What is the point of this comment?

0xdavinchee replied:
we add this everywhere we do authorizeTokenAccess as a reminder-primarily for ourselves or auditors.

0xdavinchee replied:
primarily as a reminder to ourselves and auditors, but it is not copied everywhere so will remove and add in `AgreementLibrary.sol`

kasparkallas commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -934,6 +910,120 @@ contract ConstantFlowAgreementV1 is
+    /// - passed permissions are "clean"
+    /// - no sender flow operator
+    /// - no negative allowance
+    function _validateUpdateFlowOperatorDataInput(
```
Nitpick: It's also authorizing though...

0xdavinchee replied:
using the more verbose: `_validateAndAuthorizeUpdateFlowOperatorDataInput`

kasparkallas commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -934,6 +910,120 @@ contract ConstantFlowAgreementV1 is
+        return existingPermissions | permissionDelta;
+    }
+
+    function deltaRemovePermissions(uint8 existingPermissions, uint8 permissionDelta)
```
Nitpick: IMO the "delta" prefix in the functions names is not necessary.

0xdavinchee replied:
fairly indifferent on this one, so will remove `delta`

kasparkallas commented on packages/ethereum-contracts/test/foundry/agreements/ConstantFlowAgreementV1.ACL.t.sol:
```diff
@@ -155,17 +165,138 @@ contract ConstantFlowAgreementV1ACLTest is FoundrySuperfluidTester {
+
+        superToken.increaseFlowAllowanceWithPermissions(flowOperator, permissions, flowAllowance);
+
+        uint8 newPermissions = sf.cfa.deltaAddPermissions(oldPermissions, permissions);
```
I'm not sure I like this...

Feels like testing the implementation with _the implementation_.

Is `deltaAddPermissions` covered elsewhere?

0xdavinchee replied:
we are not explicitly testing `deltaAddPermissions`.

kasparkallas commented on packages/ethereum-contracts/contracts/interfaces/agreements/IConstantFlowAgreementV1.sol:
```diff
@@ -143,6 +143,42 @@ abstract contract IConstantFlowAgreementV1 is ISuperAgreement {
+     * @param ctx Context bytes (see ISuperfluid.sol for Context struct)
+     * @return newCtx The new context bytes
+     */
+    function increaseFlowAllowanceWithPermissions(
```
It's a tiny detail but there is a bit of a new term "flow allowance" used here as opposed to "flow rate allowance".

Fun fact, in the products, we've been calling it "stream allowance".

kasparkallas replied:
Yea, some other functions are named "increaseFlow**Rate**Allowance" and "decreaseFlow**Rate**Allowance".

0xdavinchee replied:
This was done consciously, mainly due to verbosity of the function name, but I see your point here, it is inconsistent and may lead one to question if there is a difference in behavior between the two functions.

kasparkallas commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -934,6 +910,120 @@ contract ConstantFlowAgreementV1 is
+        if (newFlowRateAllowance < 0) revert CFA_ACL_NO_NEGATIVE_ALLOWANCE();
+        flowOperatorData.flowRateAllowance = newFlowRateAllowance;
+
+        token.updateAgreementData(flowOperatorId, _encodeFlowOperatorData(flowOperatorData));
```
Wouldn't it be better to use `_updateFlowRateAllowance` here or what's the point of that function?

0xdavinchee replied:
it was used previously in the cases where we modified the flow rate allowance, but kept the permissions the same (create/updateByFlowOperator and increase/decreaseAllowance) - but I will be changing the name of the function (it's fine because the function internal) to `_updateFlowOperatorData` as this is what it really does.

kasparkallas commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -826,30 +826,6 @@ contract ConstantFlowAgreementV1 is
-    }
-
     /// @dev IConstantFlowAgreementV1.increaseFlowRateAllowance implementation
     function increaseFlowRateAllowance(
```
Shouldn't this function delegate to `increaseFlowAllowanceWithPermissions` with `permissionsToAdd` set as 0?

Same goes for `decreaseFlowRateAllowance` .

Seems to me like all the code here could be de-duplicated.

0xdavinchee replied:
good catch, done

kasparkallas commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -934,6 +910,120 @@ contract ConstantFlowAgreementV1 is
+    }
+
+    function deltaAddPermissions(uint8 existingPermissions, uint8 permissionDelta)
+        public
```
Is there any value in _not_ making this public?

0xdavinchee replied:
This is public mainly for ease of testability, but given the earlier comment regarding testing the implementation with implementation, the value may lie there in not making it public.

However, testing bitwise operations with fuzzing seems a bit pedantic here.

0xdavinchee committed:
address review comments

0xdavinchee committed:
Merge branch 'dev' into 1544-new-flow-operator-updates

0xdavinchee committed:
Merge branch 'dev' into 1544-new-flow-operator-updates

hellwolf commented on packages/ethereum-contracts/testing-benchmark.json:
```diff
@@ -0,0 +1,2986 @@
+[
```
what is this file

hellwolf commented on packages/ethereum-contracts/package.json:
```diff
@@ -29,8 +29,8 @@
     "scripts": {
-        "dev": "tasks/dev.sh",
-        "dev-forge": "nodemon -e sol -x yarn run-forge test --hardhat",
+        "dev": "nodemon -e sol -x yarn run-forge test --hardhat",
```
`yarn run-forge` still and `--hardhat`?

hellwolf replied:
tweaked

hellwolf committed:
Merge branch 'dev' into 1544-new-flow-operator-updates

hellwolf committed:
Merge branch 'dev' into 1544-new-flow-operator-updates

hellwolf committed:
Merge branch 'dev' into 1544-new-flow-operator-updates

hellwolf committed:
some tweaks

hellwolf committed:
gitignore testing-benchmark.json

hellwolf committed:
run-forge
